scripts/question_answering/util.py

```diff
@@ -0,0 +1,135 @@
+r"""
+This file contains some useful function and class.
+"""
+
+import json
+import math
+
+try:
+    from config import DATA_PATH, INIT_LEARNING_RATE, WARM_UP_STEPS
+except ImportError:
+    from .config import DATA_PATH, INIT_LEARNING_RATE, WARM_UP_STEPS
+
+
+def mask_logits(x, mask):
+    r"""Implement mask logits computation.
+
+        Parameters
+        -----------
+        x : NDArray
+            input tensor with shape `(batch_size, sequence_length)`
+        mask : NDArray
+            input tensor with shape `(batch_size, sequence_length)`
+
+        Returns
+        --------
+        return : NDArray
+            output tensor with shape `(batch_size, sequence_length)`
+        """
+    return x + -1e30 * (1 - mask)
+
+
+def load_emb_mat(file_name):
+    r"""Implement load embedding matrix.
+
+        Parameters
+        -----------
+        file_name : string
+            the embedding matrix file name.
+
+        Returns
+        --------
+        mat : List[List]
+            output 2-D list.
+        """
+    with open(DATA_PATH + file_name) as f:
+        mat = json.loads(f.readline())
+    return mat
+
+
+def warm_up_lr(step):
+    r"""Implement learning rate warm up.
+
+        Parameters
+        -----------
+        step : int
+            control the learning rate linear increase.
+
+        Returns
+        --------
+        return : int
+            the learning rate for next weight update.
+        """
+    return min(INIT_LEARNING_RATE, INIT_LEARNING_RATE * (math.log(step) / math.log(WARM_UP_STEPS)))
+
+
+def zero_grad(params):
+    r"""
+    Set the grad to zero.
+    """
+    for _, paramter in params.items():
+        paramter.zero_grad()
+
+
+class ExponentialMovingAverage():

```

Comment:
This is generally very useful. We should consider moving this to MXNet/GluonNLP official API later @szha @cgraywang